### PR TITLE
call class serialize/deserialize methods

### LINKED ISSUE #19: Custom Serialize/Deserialize
What do you think about allowing classes to provide custom serialize/deserialize methods?

By adding something like this to `serializeStruct`

```ts
function serializeStruct(schema: Schema, obj: any, writer: any) {
  if (typeof obj.borshSerialize === 'function') {
    obj.borshSerialize(writer);
    return;
  }
.....
}
.....

function deserializeStruct(
  schema: Schema,
  classType: any,
  reader: BinaryReader
) {
  if(typeof classType.borshDeserialize) {
    classType.borshDeserialize(reader);
    return;
  }
....
}
```

we could then do this:

```ts
export class PublicKey {
    keyType: KeyType;
    data: Uint8Array;

    static borsheDeserialize(reader: BinaryReader) {
        reader.readU8
    }

    static borshDeserialize(reader: BinaryReader): PublicKey {
        const keyType = reader.readU8();

        let data;
        if(keyType === KeyType.ED25519) {
            data = reader.readFixedArray(32);
        } else if(keyType === KeyType.SECP256k1) {
            data = reader.readFixedArray(64);
        }

        return new PublicKey({ keyType, data });
    }

    borshSerialize(writer: BinaryWriter) {
        writer.writeU8(this.keyType);
        writer.writeFixedArray(this.data);
    }
}
```

If this looks good I'm happy to implement it and submit a PR.

### PR DESCRIPTION
**Pull request by mehtaphysical**
Fixes #19 

This allows for something like:

```ts
export class PublicKey extends Assignable {
    keyType: KeyType;
    data: Uint8Array;

    serialize(writer: BinaryWriter) {
        writer.writeU8(this.keyType);
        writer.writeFixedArray(this.data);
    }

    static deserialize(reader: BinaryReader): PublicKey {
        const keyType = reader.readU8();
        const size = KeyType.ED25519 ? 32 : 65;
        return new PublicKey({ keyType, data: reader.readFixedArray(size) });
    }
}
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -45,8 +46,7 @@
     "typescript": "^3.6.2"
   },
   "dependencies": {
-    "@types/bn.js": "^4.11.5",
-    "bn.js": "^5.0.0",
+    "bn.js": "^5.2.0",
```
Comment (ailisp) [inline package.json:49]: nice cleanup!

File: lib/index.js [hunk 2]
```diff
@@ -291,6 +291,10 @@ function serializeField(schema, fieldName, value, fieldType, writer) {
     }
 }
 function serializeStruct(schema, obj, writer) {
+    if (typeof obj.serialize === 'function') {
+        obj.serialize(writer);
```
Comment (ailisp) [inline lib/index.js:295]: I would prefer obj.borshSerialize/obj.borshDeserialize for a less chance of name collision

**Review by ailisp — APPROVED** [review]
Code change looks good to me, @volovyk-s do you think the API change makes sense?

**Review by volovyk-s — CHANGES_REQUESTED** [review]
Looks good overall. Let's add simple unit tests and merge this PR.

**Review by volovyk-s — APPROVED** [review]

**Pull request merged**